### ISSUE TITLE
Query could return wrong result if composite index defined on a class

### ISSUE THREAD

**Issue by Kikimora** [orientechnologies/orientdb#5536]
I have Person class and index created like this
```
person.createIndex("Person_name_birthday_btree", OClass.INDEX_TYPE.NOTUNIQUE, "name", "birthday");
```
I also have two persons in DB. One with name 'Shah Rukh Khan' and birthday != null another with name = 'Lakshman .' and birthday == null

Then I issue query
```
select from Person where name = 'Shah Rukh Khan'
```
I get my person back.
But when i query
```
select from Person where name = 'Lakshman .'
```
I get null as a result.

My workaround for this is create index with `METADATA { ignoreNullValues : false }`

**Comment by luigidellaquila** [comment]
IMHO the settings `METADATA { ignoreNullValues : false }` should be the default, or even better, indexes with `ignoreNullValues = false` should never be used to optimize queries (eg. they should be used only for direct access with `SELECT FORM index:...`).

This change will break the performance all the existing databases, almost all the existing indexes will be useless.

Probably this can be put in the roadmap for 3.0, with proper documentation and with migration guide

WDYT?

**Comment by laa** [comment]
Hi Luigi,
I do not support it, because insertion with index which support null keys
for values may be much slower than in standard index.
So I suggest to leave everything as is.

On Thu, Mar 17, 2016 at 3:41 PM luigidellaquila <notifications@github.com>
wrote:

> IMHO the settings METADATA { ignoreNullValues : false } should be the
> default, or even better, indexes with ignoreNullValues = false should
> never be used to optimize queries (eg. they should be used only for direct
> access with SELECT FORM index:...).
>
> This change will break the performance all the existing databases, almost
> all the existing indexes will be useless.
>
> Probably this can be put in the roadmap for 3.0, with proper documentation
> and with migration guide
>
> WDYT?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/orientechnologies/orientdb/issues/5536#issuecomment-197883702>
>
-- 
Best regards,
Andrey Lomakin, R&D lead.
OrientDB Ltd

twitter:@Andrey_Lomakin linkedin:https://ua.linkedin.com/in/andreylomakin

**Comment by luigidellaquila** [comment]
Hi @laa 

so which is the solution to have correct results by default?

Thanks

Luigi

**Comment by laa** [comment]
Well as I can see we all agree to support null values by default.

On Tue, Mar 29, 2016 at 11:59 AM luigidellaquila <notifications@github.com>
wrote:

> Hi @laa <https://github.com/laa>
>
> so which is the solution to have correct results by default?
>
> Thanks
>
> Luigi
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/orientechnologies/orientdb/issues/5536#issuecomment-202785634>
>
-- 
Best regards,
Andrey Lomakin, R&D lead.
OrientDB Ltd

twitter:@Andrey_Lomakin linkedin:https://ua.linkedin.com/in/andreylomakin

**Comment by luigidellaquila** [comment]
Hi @Kikimora 

This was implemented in 2.2 GA, now all the indexes are created with `METADATA { ignoreNullValues : false }` by default.

Moreover, indexes that ignore null values are ignored on partial index matching in queries

Closing

Thanks

Luigi
